File: src/common/tcpdump_manager.py
import os
import subprocess
from datetime import datetime

from common.logger import log

logger = log("TcpDumpManager")


class TcpDumpManager:
    """
    Manages the execution of tcpdump for packet capture.
    """

    def __init__(
        self,
        interface: str,
        output_path: str,
        target_host: str,
        enabled: bool = False,
    ):
        """
        Initializes the TcpDumpManager.

        Args:
            interface (str): Network interface to capture on.
            output_path (str): Directory to save pcap files.
            target_host (str): IP or hostname to filter by.
            enabled (bool): Whether packet capture is enabled.
        """
        self.interface = interface
        self.output_path = output_path
        self.target_host = target_host
        self.enabled = enabled
        self.process = None
        self.capture_file = None

    def __enter__(self):
        self.start()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.stop()

    def start(self):
        """
        Starts the tcpdump process in the background.
        """
        if not self.enabled:
            return

        if not os.path.exists(self.output_path):
            try:
                os.makedirs(self.output_path)
            except OSError as e:
                logger.error(
                    f"Failed to create output directory {self.output_path}: "
                    f"{e}"
                )
                return

        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        self.capture_file = os.path.join(
            self.output_path, f"experiment_capture_{timestamp}.pcap"
        )

        # Construct the command
        # -i: Interface
        # -w: Write to file
        # host: Filter by host
        command = [
            "tcpdump",
            "-i",
            self.interface,
            "-w",
            self.capture_file,
            "host",
            self.target_host,
        ]

        try:
            logger.info(f"Starting tcpdump with command: {' '.join(command)}")
            self.process = subprocess.Popen(
                command, stdout=subprocess.PIPE, stderr=subprocess.PIPE
            )
            logger.info(
                f"Packet capture started. Saving to: {self.capture_file}"
            )
        except Exception as e:
            logger.error(f"Failed to start tcpdump: {e}")
            self.process = None

    def stop(self):
        """
        Stops the tcpdump process.
        """
        if not self.enabled or self.process is None:
            return

        logger.info("Stopping tcpdump...")
        try:
            self.process.terminate()
            try:
                self.process.wait(timeout=5)
            except subprocess.TimeoutExpired:
                logger.warning(
                    "tcpdump did not terminate gracefully, killing..."
                )
                self.process.kill()
                self.process.wait()

            logger.info("tcpdump stopped successfully.")

            if os.path.exists(self.capture_file):
                size = os.path.getsize(self.capture_file)
                logger.info(
                    f"Capture file created: {self.capture_file} ({size} bytes)"
                )
            else:
                logger.warning(f"Capture file not found: {self.capture_file}")

        except Exception as e:
            logger.error(f"Error stopping tcpdump: {e}")
        finally:
            self.process = None

File: rust_producer/src/tcpdump_manager.rs
use std::fs;
use std::path::Path;
use std::process::{Child, Command, Stdio};
use chrono::Local;
use log::{error, info};
use crate::config::Config;

pub struct TcpDumpManager {
    enabled: bool,
    interface: String,
    output_path: String,
    target_host: String,
    process: Option<Child>,
    capture_file: Option<String>,
}

impl TcpDumpManager {
    pub fn new(config: &Config) -> Self {
        let target_host = config.target_host.clone().unwrap_or_else(|| {
            config.bootstrap_servers
                .split(':')
                .next()
                .unwrap_or("localhost")
                .to_string()
        });

        Self {
            enabled: config.tcpdump_enabled,
            interface: config.tcpdump_interface.clone(),
            output_path: config.tcpdump_output_path.clone(),
            target_host,
            process: None,
            capture_file: None,
        }
    }

    pub fn start(&mut self) {
        if !self.enabled {
            return;
        }

        if !Path::new(&self.output_path).exists() {
            match fs::create_dir_all(&self.output_path) {
                Ok(_) => {},
                Err(e) => {
                    error!("Failed to create output directory {}: {}", self.output_path, e);
                    return;
                }
            }
        }

        let timestamp = Local::now().format("%Y%m%d_%H%M%S");
        let capture_file = format!("{}/experiment_capture_{}.pcap", self.output_path, timestamp);
        self.capture_file = Some(capture_file.clone());

        info!("Starting tcpdump on interface {} filtering host {}", self.interface, self.target_host);

        match Command::new("tcpdump")
            .arg("-i")
            .arg(&self.interface)
            .arg("-w")
            .arg(&capture_file)
            .arg("host")
            .arg(&self.target_host)
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()
        {
            Ok(child) => {
                info!("Packet capture started. Saving to: {}", capture_file);
                self.process = Some(child);
            }
            Err(e) => {
                error!("Failed to start tcpdump: {}", e);
                self.process = None;
            }
        }
    }

    pub fn stop(&mut self) {
        if !self.enabled {
            return;
        }

        if let Some(mut child) = self.process.take() {
            info!("Stopping tcpdump...");
            match child.kill() {
                Ok(_) => {
                    match child.wait() {
                        Ok(_) => info!("tcpdump stopped successfully."),
                        Err(e) => error!("Failed to wait for tcpdump process: {}", e),
                    }
                }
                Err(e) => error!("Failed to kill tcpdump process: {}", e),
            }
        }
    }
}

impl Drop for TcpDumpManager {
    fn drop(&mut self) {
        self.stop();
    }
}
